docs/config.md
# Config doc

Define the values of config as default in [config.ts](../config.ts).
These properties in details are as follows.

| Large Category | Small Category | Property               | Description                                                                                                               | Type                                         | Default value                                                                                 |
| -------------- | -------------- | ---------------------- | ------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------- | --------------------------------------------------------------------------------------------- |
| base           |                |                        | Define base configuration                                                                                                 |
|                |                | alias                  | Define the identifying your stack                                                                                         | `string`                                     |                                                                                               |
| networkConfig  |                |                        | Define the network configuration                                                                                          | `NetworkConfig`                              |
|                |                | cidr                   | Vpc CIDR of multitenancy and each corp network                                                                            | `string`                                     | `10.0.0.0/16`                                                                                 |
|                |                | cidrMask               | Define the CIDR of `publicSubnet`,`natSubnet` and `isolatedSubnet`                                                        | `number`                                     | `24`                                                                                          |
|                |                | publicSubnet           | Define whether creating a public subnet or not                                                                            | `boolean`                                    | `false`                                                                                       |
|                |                | natSubnet              | Define whether creating a nat subnet (a private subnet with NAT gateway) or not                                           | `boolean`                                    | `false`                                                                                       |
|                |                | isolatedSubnet         | Define whether creating a isolated subnet (a private subnet without NAT gateway) or not                                   | `boolean`                                    | `false`                                                                                       |
|                |                | maxAzs                 | Define how many AZs in the region are created                                                                             | `number`                                     | `2`                                                                                           |
|                |                | zoneName               | Define the domain name of PrivatedHostedZone in Route53                                                                   | `string`                                     |                                                                                               |
|                |                | adEnv                  | Define whether creating some vpc endpoints or not                                                                         | `boolean`                                    | `false`                                                                                       |
| keyCloakApp    |                |                        | Define KeyCloak configuration                                                                                             | `KeyCloakAppConfig`                          |
|                | dbConfig       | engine                 | Define the Aurora Serverless v2 cluster engine                                                                            | `IClusterEngine`                             | `AuroraPostgresEngineVersion.VER_15_2`                                                        |
|                | dbConfig       | v2MaxCapacity          | Define the minimum number of Aurora capacity units (ACUs) for a DB instance in an Aurora Serverless v2 cluster            | `number`                                     | `32`                                                                                          |
|                | dbConfig       | v2MinCapacity          | Define the minimum number of Aurora capacity units (ACUs) for a DB instance in an Aurora Serverless v2 cluster            | `number`                                     | `0.5`                                                                                         |
|                | dbConfig       | metricDuration         | Define how frequency caliculating of the database metrics                                                                 | `Duration`                                   | `Duration.minutes(10)`                                                                        |
|                | dbConfig       | databaseCapacityOption | Define the alarm configuration of the average of the `ServerlessDatabaseCapacity` values                                  | `CreateAlarmOptions`                         | `threshold: 1.5`,`evaluationPeriods: 3`                                                       |
|                | dbConfig       | acuUtilOption          | Define the alarm configuration of the average of the `ServerlessDatabaseCapacity` metric divided by the maximum ACU value | `CreateAlarmOptions`                         | `threshold: 90`, `evaluationPeriods: 3`                                                       |
|                | dbConfig       | dbUserName             | Define the administrative username of the database                                                                        | `string`                                     | `awsdemo`                                                                                     |
|                | dbConfig       | dbname                 | Define the name of the database                                                                                           | `string`                                     | `awssample`                                                                                   |
|                | dbConfig       | port                   | Define the database listen port                                                                                           | `number`                                     | `5432`                                                                                        |
|                | dbConfig       | parameterName          | Define the parameter name in SSM parameter store                                                                          | `string`                                     | `keycloak`                                                                                    |
|                | dbConfig       | enableProxy            | Define whether creating the rds proxy or not                                                                              | `boolean`                                    | `false`                                                                                       |
|                | appConfig      | acmCertificateArn      | Define the arn of acm for keycloak                                                                                        | `AcmCertificateArn`                          | `service:string`,`keycloackForAdmin:string`                                                   |
|                | appConfig      | fargateTaskDefinition  | Define the details of a task definition run on a Fargate cluster                                                          | `FargateTaskDefinitionProps`                 | `cpu: 1024`, `memoryLimitMiB: 2048`                                                           |
|                | appConfig      | fargateService         | Define the properties for the ApplicationLoadBalancedFargateService service                                               | `ApplicationLoadBalancedFargateServiceProps` | `desiredCount: 1`,`enableExecuteCommand: true`                                                |
|                | appConfig      | runtimePlatform        | Define the Runtime Platform in a fargate                                                                                  | `RuntimePlatform`                            | `operatingSystemFamily: OperatingSystemFamily.LINUX`,`cpuArchitecture: CpuArchitecture.ARM64` |
|                | appConfig      | nlbConfig              | Define the properties of LB                                                                                               | `LbConfig`                                   | `listenerPort: 443`,`healthCheckPath: '/health'`,`targetPort: 8080`                           |
|                | appConfig      | portMapping            | Define the port mappings to add to the container definition                                                               | `PortMapping[]`                              | `{hostPort: 8080,containerPort: 8080,},{hostPort: 7800,containerPort: 7800}`                  |
|                | appConfig      | imagePath              | Define the image path for pushing it in ECR                                                                               | `string`                                     | `path.join(__dirname, './', 'docker/keycloak')`                                               |
| nuxtApp        |                |                        | Define Nuxt3 App configuration                                                                                            | `NuxtAppConfig`                              |
|                | corpList       | name                   | Define the corp name                                                                                                      | `string`                                     |
|                | corpList       | dedicated              | Define whether creating the dedicated NLB for the corp                                                                    | `boolean`                                    |
|                | corpList       | acmCertificateArn      | Define the arn of acm for the dedicated NLB if the `dedicated` is true                                                    | `string`                                     |                                                                                               |
|                | corpList       | overrideDbConfig       | Define the properties overriding the default DB config                                                                    | `OverridDbConfig`                            |
|                | dbConfig       | engine                 | Define the Aurora Serverless v2 cluster engine                                                                            | `IClusterEngine`                             | `AuroraPostgresEngineVersion.VER_15_2`                                                        |
|                | dbConfig       | v2MaxCapacity          | Define the maximum number of Aurora capacity units (ACUs) for a DB instance in an Aurora Serverless v2 cluster            | `number`                                     | `32`                                                                                          |
|                | dbConfig       | v2MinCapacity          | Define the minimum number of Aurora capacity units (ACUs) for a DB instance in an Aurora Serverless v2 cluster            | `number`                                     | `0.5`                                                                                         |
|                | dbConfig       | metricDuration         | Define how frequency caliculating of the database metrics                                                                 | `Duration`                                   | `Duration.minutes(10) `                                                                       |
|                | dbConfig       | databaseCapacityOption | Define the alarm configuration of the average of the `ServerlessDatabaseCapacity` values                                  | `CreateAlarmOptions`                         | `threshold: 1.5`,`evaluationPeriods: 3`                                                       |
|                | dbConfig       | acuUtilOption          | Define the alarm configuration of the average of the `ServerlessDatabaseCapacity` metric divided by the maximum ACU value | `CreateAlarmOptions`                         | `threshold: 90`, `evaluationPeriods: 3`                                                       |
|                | dbConfig       | dbUserName             | Define the administrative username of the database                                                                        | `string`                                     | `corpdemo`                                                                                    |
|                | dbConfig       | dbname                 | Define the name of the database                                                                                           | `string`                                     | `corpsample`                                                                                  |
|                | dbConfig       | port                   | Define the database listen port                                                                                           | `number`                                     | `5432`                                                                                        |
|                | dbConfig       | parameterName          | Define the parameter name in SSM parameter store                                                                          | `string`                                     | `corp`                                                                                        |
|                | dbConfig       | enableProxy            | Define whether creating the rds proxy or not                                                                              | `boolean`                                    | `true`                                                                                        |
|                | appConfig      | acmCertificateArn      | Define the arn of acm for Nuxt3 app                                                                                       | `AcmCertificateArn`                          | `service:string`                                                                              |
|                | appConfig      | fargateTaskDefinition  | Define the details of a task definition run on a Fargate cluster                                                          | `FargateTaskDefinitionProps`                 | `cpu: 1024`, `memoryLimitMiB: 2048`,                                                          |
|                | appConfig      | fargateService         | Define the properties for the ApplicationLoadBalancedFargateService service                                               | `ApplicationLoadBalancedFargateServiceProps` | `desiredCount: 1`,`enableExecuteCommand: true`                                                |
|                | appConfig      | runtimePlatform        | Define the Runtime Platform in a fargate                                                                                  | `RuntimePlatform`                            | `operatingSystemFamily: OperatingSystemFamily.LINUX`,`cpuArchitecture: CpuArchitecture.ARM64` |
|                | appConfig      | nlbConfig              | Define the properties of LB                                                                                               | `LbConfig`                                   | `listenerPort: 443`,`healthCheckPath: '/health'`,`targetPort: 3000`                           |
|                | appConfig      | portMapping            | Define the port mappings to add to the container definition                                                               | `PortMapping[]`                              | `{hostPort: 3000,containerPort: 3000}`                                                        |
|                | appConfig      | imagePath              | Define the image path for pushing it in ECR                                                                               | `string`                                     | `imagePath: path.join(__dirname, './', 'docker/nuxtjs')`                                      |


config.ts
import { AuroraPostgresEngineVersion, DatabaseClusterEngine } from 'aws-cdk-lib/aws-rds';
import { Config } from './config.types';
import { Duration } from 'aws-cdk-lib';
import { CpuArchitecture, OperatingSystemFamily } from 'aws-cdk-lib/aws-ecs';
import path = require('path');
export const devConfig: Config = {
  base: {
    alias: 'hkoizum',
  },
  networkConfig: {
    cidr: '10.0.0.0/16',
    cidrMask: 24,
    publicSubnet: false,
    natSubnet: false,
    isolatedSubnet: true,
    maxAzs: 2,
    zoneName: 'multitenancy.com',
    adEnv: false,
  },
  keyCloakApp: {
    dbConfig: {
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_15_2,
      }),
      v2MaxCapacity: 32,
      v2MinCapacity: 0.5,
      metricDuration: Duration.minutes(10),
      databaseCapacityOption: {
        threshold: 1.5,
        evaluationPeriods: 3,
      },
      acuUtilOption: {
        threshold: 90,
        evaluationPeriods: 3,
      },
      dbUserName: 'awsdemo',
      dbname: 'awssample',
      port: 5432,
      parameterName: 'keycloak',
      enableProxy: false,
    },
    appConfig: {
      acmCertificateArn: {
        service:
          'arn:aws:acm:ap-northeast-1:123456789012:certificate/8980a069-827b-4c74-9833-9522c3f32962',
        keycloackForAdmin:
          'arn:aws:acm:ap-northeast-1:123456789012:certificate/914aa0f1-ff3e-4b30-9300-7345dad453a7',
      },
      fargateTaskDefinition: {
        cpu: 1024,
        memoryLimitMiB: 2048,
      },
      fargateService: {
        desiredCount: 1,
        enableExecuteCommand: true,
      },
      runtimePlatform: {
        operatingSystemFamily: OperatingSystemFamily.LINUX,
        cpuArchitecture: CpuArchitecture.ARM64,
      },
      nlbConfig: {
        listenerPort: 443,
        healthCheckPath: '/health',
        targetPort: 8080,
      },
      portMapping: [
        {
          hostPort: 8080,
          containerPort: 8080,
        },
        {
          hostPort: 7800,
          containerPort: 7800,
        },
      ],
      imagePath: path.join(__dirname, './', 'docker/keycloak'),
    },
  },
  nuxtApp: {
    corpList: [
      {
        name: 'a',
        dedicated: false,
        overrideDbConfig: {
          dbUserName: 'acorpdemo',
          dbname: 'acorpsample',
          parameterName: 'a-corp',
        },
      },
      {
        name: 'b',
        dedicated: false,
        overrideDbConfig: {
          dbUserName: 'bcorpdemo',
          dbname: 'bcorpsample',
          parameterName: 'b-corp',
        },
      },
      {
        name: 'c',
        dedicated: true,
        acmCertificateArn:
          'arn:aws:acm:ap-northeast-1:123456789012:certificate/815d888f-89c9-4411-b8f5-d790a187c18f',
        overrideDbConfig: {
          dbUserName: 'ccorpdemo',
          dbname: 'ccorpsample',
          parameterName: 'c-corp',
        },
      },
    ],

    dbConfig: {
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_15_2,
      }),
      v2MaxCapacity: 32,
      v2MinCapacity: 0.5,
      metricDuration: Duration.minutes(10),
      databaseCapacityOption: {
        threshold: 1.5,
        evaluationPeriods: 3,
      },
      acuUtilOption: {
        threshold: 90,
        evaluationPeriods: 3,
      },
      dbUserName: 'corpdemo',
      dbname: 'corpsample',
      port: 5432,
      parameterName: 'corp',
      enableProxy: true,
    },
    appConfig: {
      acmCertificateArn: {
        service:
          'arn:aws:acm:ap-northeast-1:123456789012:certificate/dc2cf171-3a0a-48ef-a3aa-458bf2c375de',
      },
      fargateTaskDefinition: {
        cpu: 1024,
        memoryLimitMiB: 2048,
      },
      fargateService: {
        desiredCount: 1,
        enableExecuteCommand: true,
      },
      runtimePlatform: {
        operatingSystemFamily: OperatingSystemFamily.LINUX,
        cpuArchitecture: CpuArchitecture.ARM64,
      },
      nlbConfig: {
        listenerPort: 443,
        healthCheckPath: '/',
        targetPort: 3000,
      },
      portMapping: [
        {
          hostPort: 3000,
          containerPort: 3000,
        },
      ],
      imagePath: path.join(__dirname, './', 'docker/nuxtjs'),
    },
  },
};
